nexus/server.go

```diff
@@ -0,0 +1,497 @@
+		operationState = unsuccessfulError.State
+		failure = &unsuccessfulError.Failure
+		statusCode = statusOperationFailed
+
+		if operationState == OperationStateFailed || operationState == OperationStateCanceled {
+			writer.Header().Set(headerOperationState, string(operationState))
+		} else {
+			h.logger.Error("unexpected operation state", "state", operationState)
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	} else if errors.As(err, &handlerError) {
+		failure = handlerError.Failure
+		statusCode = handlerError.StatusCode
+	} else {
+		failure = &Failure{
+			Message: "internal server error",
+		}
+		h.logger.Error("handler failed", "error", err)
+	}
+
+	var bytes []byte
+	if failure != nil {
+		bytes, err = json.Marshal(failure)
+		if err != nil {
+			h.logger.Error("failed to marshal failure", "error", err)
+			writer.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		writer.Header().Set(headerContentType, contentTypeJSON)
+	}
+
+	writer.WriteHeader(statusCode)
+
+	if _, err := writer.Write(bytes); err != nil {
+		h.logger.Error("failed to write response body", "error", err)
+	}
+}
+
+func (h *httpHandler) startOperation(writer http.ResponseWriter, request *http.Request) {
+	operation, err := url.PathUnescape(path.Base(request.URL.RawPath))
+	if err != nil {
+		h.writeFailure(writer, newBadRequestError("failed to parse URL path"))
+		return
+	}
+	handlerRequest := &StartOperationRequest{
+		Operation:   operation,
+		RequestID:   request.Header.Get(headerRequestID),
+		CallbackURL: request.URL.Query().Get(queryCallbackURL),
+		HTTPRequest: request,
+	}
+	response, err := h.options.Handler.StartOperation(request.Context(), handlerRequest)
+	if err != nil {
+		h.writeFailure(writer, err)
+	} else {
+		response.applyToHTTPResponse(writer, h)
+	}
+}
+
+func (h *httpHandler) getOperationResult(writer http.ResponseWriter, request *http.Request) {
+	// strip /result
+	prefix, operationIDEscaped := path.Split(path.Dir(request.URL.RawPath))
+	operationID, err := url.PathUnescape(operationIDEscaped)
+	if err != nil {
+		h.writeFailure(writer, newBadRequestError("failed to parse URL path"))
+		return
+	}
+	operation, err := url.PathUnescape(path.Base(prefix))
+	if err != nil {
+		h.writeFailure(writer, newBadRequestError("failed to parse URL path"))
+		return
+	}
+	handlerRequest := &GetOperationResultRequest{Operation: operation, OperationID: operationID, HTTPRequest: request}
+
+	waitStr := request.URL.Query().Get(queryWait)
+	ctx := request.Context()
+	if waitStr != "" {
+		waitDuration, err := time.ParseDuration(waitStr)
+		if err != nil {
+			h.logger.Warn("invalid wait duration query parameter", "wait", waitStr)
+			h.writeFailure(writer, newBadRequestError("invalid wait query parameter"))
+			return
+		}
+		handlerRequest.Wait = waitDuration
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(request.Context(), h.options.GetResultTimeout)
+		defer cancel()
```

Comment:
I think this context timeout should be created regardless of wait value. Hopefully doesn't matter since good handlers should not wait if there is no wait, but setting the context timeout to the get result timeout is good practice always.

Reply:
This should be applied to all endpoints then, not just get-result.

Reply:
Yeah, can stay in this conditional